### ISSUE TITLE
(react/jsx-key) false positive for a not array .map() method

### ISSUE THREAD

**Issue by scink** [yannickcr/eslint-plugin-react#2487]
i use library [fp-ts](https://github.com/gcanti/fp-ts)@1 and there is, for example, an Option. it has a method map, but it has nothing common with an array type. there is no need in `key` but eslint thinks differently. i think, it just search for `.map(` in jsx, but it has to check the type.
```
const qwe = {
    map: (a: () => ReactNode) => a(),
};
const asd = () => (
    <span>
        {qwe.map(() => (
            <span>qwe</span>
        ))}
    </span>
);
```
![Screenshot 2019-11-05 at 18 57 23](https://user-images.githubusercontent.com/15276485/68223604-2344d380-fffe-11e9-8160-32623e5b69e2.png)

and it does not fail on this, thought it should
```
const map = <A, B>(m: (a: A) => B) => (as: A[]): B[] => as.map(m);
const asd = () => <span>{map(() => <span>qwe</span>)([])}</span>;
```
![Screenshot 2019-11-05 at 19 00 28](https://user-images.githubusercontent.com/15276485/68223888-b1b95500-fffe-11e9-82f3-96a8458fc2fe.png)

**Comment by golopot** [comment]
It is hard to find out the type of a variable in ESLint, so the current cheap solution is to blindly assume that every `.map` is the array `.map`. In the future it might be possible to get type informations might with  [@typescript-eslint/experimental-utils
](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/experimental-utils) .

**Comment by ljharb** [comment]
The solution here is to disable this rule, either overall or with an override comment. There's no way to reliably - for JavaScript - determine if a variable is an array or not, so "anything with an array method name" is treated as one.

**Comment by OliverJAsh** [comment]
Similar story here—we're using an `Option` type, along with other types that also implement a `map` method, such as `RemoteData`/`Observable`/`Either`/`Validation`.

I looked into enabling this rule on our project and it produced 68 false positives. I don't think it would be reasonable for us to add disable comments in that many places.

In any case, I realise our use case is probably the exception, not the rule!

**Comment by SRachamim** [comment]
I suggest to reopen this issue and fix it by inspecting the real type. Assuming that anything that has a `map` key is an array is like assuming that anything that has a `then` key is a promise 😋.

Instead of guessing the types of entities in the code, let's really inspect it. See this rule as an example: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/strict-boolean-expressions.ts
